application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbLocalSubscriptionService.java

```diff
@@ -343,15 +343,29 @@ private void onTimeSeriesUpdate(UUID entityId, List<TsKvEntry> data, TbCallback
                 s -> {
                     TbTimeSeriesSubscription sub = (TbTimeSeriesSubscription) s;
                     List<TsKvEntry> updateData = null;
+                    Map<String, Long> keyStates = sub.getKeyStates();
                     if (sub.isAllKeys()) {
-                        updateData = data;
+                        if (sub.isLatestValues()) {
+                            for (TsKvEntry kv : data) {
+                                if (!keyStates.containsKey((kv.getKey())) || kv.getTs() > keyStates.get(kv.getKey())) {
```

Comment:
Using `keyStates.get(kv.getKey())` and `keyStates.containsKey((kv.getKey())` executed the hash map search twice. Refactor this to search in the map by key only once.